src/main/java/io/confluent/idesidecar/websocket/resources/WebsocketEndpoint.java

```diff
@@ -70,17 +146,30 @@ public void broadcast(Message message) throws IOException, IllegalArgumentExcept
         headers.id()
     );
 
-    // And broadcast the message
-    sessions.entrySet().stream()
-        .filter(pair -> pair.getKey().isOpen())
-        .forEach(pair -> {
-          Log.debugf(
-              "Sending broadcast message %s to workspace: %d",
-              headers.id(),
-              pair.getValue().processId()
-          );
-          pair.getKey().getAsyncRemote().sendText(jsonMessage);
-        });
+
+    // And broadcast the message to the active sessions, collecting the futures along the way.
+    List<Future> futures = new ArrayList<>();
+    activeWorkspaceSessions.forEach(ws -> {
+      Log.debugf(
+          "Sending broadcast message %s to workspace pid %d",
+          headers.id(),
+          ws.workspacePid().id()
+      );
+      futures.add(ws.session().getAsyncRemote().sendText(jsonMessage));
+    });
+
+    // Wait for all the futures to complete.
+    for (Future f : futures) {
+      try {
+        f.get();
+      } catch (Exception e) {
+        Log.errorf("Failed to send broadcast message to workspace: %s", e.getMessage());
+
+        if (e instanceof InterruptedException) {
+          Thread.currentThread().interrupt();
+        }
+      }
+    }
```

Comment:
We don't want to block here. First, it might block forever and block the thread. Second, it's really an anti-pattern in Quarkus.

The proper way to do this is to return some sort of promise, which in Quarkus is `Uni` for single objects or `Multi` for multiple items. So the first step is to add a helper method to send a message asynchronously in `WebsocketWorkspaceSession`:
```
    /**
     * Send the text message with the given message ID, and return a {@link Uni} that returns
     * the message ID when the message is sent.
     *
     * @param messageText
     * @param messageId
     * @return the message ID when the message is sent.
     */
    public Uni<String> sendAsync(String messageText, String messageId) {
      Log.debugf(
          "Sending broadcast message %s to workspace pid %d",
          messageId,
          workspacePid.id()
      );

      var future = CompletableFuture.supplyAsync(() -> {
        session.getAsyncRemote().sendText(messageText, result -> {
          if (result.isOK()) {
            Log.debugf(
                "Successfully sent broadcast message %s to workspace pid %d",
                messageId,
                workspacePid.id()
            );
          } else {
            var cause = result.getException();
            Log.errorf("Failed to send broadcast message to workspace: %s", cause.getMessage());
            throw new CompletionException(cause);
          }
        });
        return messageId;
      });
      return Uni.createFrom().completionStage(future);
    }
```
Then we can have multiple forms of `broadcast`:
```
  /**
   * Broadcast a message originating from sidecar to all active connected workspaces.
   *
   * @param message the message to broadcast
   * @throws IOException              if there is an error serializing the message to JSON.
   * @throws IllegalArgumentException if the headers of the message are not valid for broadcasting.
   * */
  public Multi<String> broadcast(Message message) throws IOException, IllegalArgumentException {
    return broadcast(message, WorkspaceWebsocketSession::isActive, "active");
  }

  /**
   * Broadcast a message originating from sidecar to connected workspaces that match the supplied
   * filter.
   *
   * @param message           the message to broadcast
   * @param filter            the filter to apply to the workspace sessions
   * @param sessionsAdjective the adjective to describe the sessions in the log message.
   * @throws IOException              if there is an error serializing the message to JSON.
   * @throws IllegalArgumentException if the headers of the message are not valid for broadcasting.
   * */
  public Multi<String> broadcast(
      Message message,
      Predicate<WorkspaceWebsocketSession> filter,
      String sessionsAdjective
  ) throws IOException, IllegalArgumentException {
    var activeWorkspaceSessions = sessions.values().stream()
                                          .filter(filter)
                                          .toList();

    if (activeWorkspaceSessions.isEmpty()) {
      Log.debugf("No %s workspaces to broadcast message to.", sessionsAdjective);
      return Multi.createFrom().empty();
    }

    final MessageHeaders headers = validateHeadersForSidecarBroadcast(message);

    // Serialize the message to JSON
    String jsonMessage = mapper.writeValueAsString(message);
    Log.debugf(
        "Broadcasting %d char message, id %s to all workspaces",
        jsonMessage.length(),
        headers.id()
    );

    return broadcast(jsonMessage, headers.id(), filter, sessionsAdjective);
  }

  /**
   * Broadcast a message originating from sidecar to connected workspaces that match the supplied
   * filter.
   *
   * @param jsonMessage       the message to broadcast
   * @param messageId         the ID of the message (used for logging)
   * @param filter            the filter to apply to the workspace sessions
   * @param sessionsAdjective the adjective to describe the sessions in the log message.
   * @throws IOException              if there is an error serializing the message to JSON.
   * @throws IllegalArgumentException if the headers of the message are not valid for broadcasting.
   * */
  public Multi<String> broadcast(
      String jsonMessage,
      String messageId,
      Predicate<WorkspaceWebsocketSession> filter,
      String sessionsAdjective
  ) {
    // Compute the sessions that satisfy the filter
    var activeWorkspaceSessions = sessions.values().stream()
                                          .filter(filter)
                                          .toList();
    if (activeWorkspaceSessions.isEmpty()) {
      Log.debugf("No %s workspaces to broadcast message to.", sessionsAdjective);
      return Multi.createFrom().empty();
    }

    var promises = activeWorkspaceSessions
        .stream()
        .map(session -> session.sendAsync(jsonMessage, messageId))
        .toList();
    return Multi
        .createFrom()
        .iterable(promises)
        .onItem()
        .transformToMultiAndMerge(Uni::toMulti);
  }
```
If a caller doesn't care about waiting, they can ignore the `Multi` response. But if they do, they can do [various things](https://smallrye.io/smallrye-mutiny/latest/guides/imperative-to-reactive/), including [handling errors](https://smallrye.io/smallrye-mutiny/latest/tutorials/handling-failures/).

For example, a test might want to block until the broadcast is completed:
```
    endpoints.broadcast(...).collect().asList().await().atMost(Duration.ofSeconds(5));
```

Reply:
Accepted your side-branch, thank you.